src/init/starship.ps1

```diff
@@ -106,7 +106,13 @@ $null = New-Module starship {
         $arguments += "--status=$($lastExitCodeForPrompt)"
 
         # Invoke Starship
-        Invoke-Native -Executable ::STARSHIP:: -Arguments $arguments
+        $prompt = Invoke-Native -Executable ::STARSHIP:: -Arguments $arguments
+
+        # Set the number of extra lines in the prompt for PSReadLine prompt redraw.
+        Set-PSReadLineOption -ExtraPromptLineCount ($prompt.Split("`n").Length - 1)
```

Comment:
If I understand correctly, this will be called on prompt, right? How much time does it take?

Reply:
Yes. It will. Oh My Posh does this as well https://github.com/JanDeDobbeleer/oh-my-posh/blob/ad5c77a61b95cc71d8864d259cefe3e54a98711b/src/init/omp.ps1#L96. Minus the `Split` since it has an array of lines.

What would be the best way to measure the difference?

Reply:
Hmm, I'm not sure. Maybe capture this statement in a `Measure-Command` and append the time to a file?

Reply:
I did this:
```diff
diff --git a/src/init/starship.ps1 b/src/init/starship.ps1
index 79ce6800..3eff46bc 100755
--- a/src/init/starship.ps1
+++ b/src/init/starship.ps1
@@ -109,7 +109,7 @@ $null = New-Module starship {
         $prompt = Invoke-Native -Executable ::STARSHIP:: -Arguments $arguments
 
         # Set the number of extra lines in the prompt for PSReadLine prompt redraw.
-        Set-PSReadLineOption -ExtraPromptLineCount ($prompt.Split("`n").Length - 1)
+        Measure-Command { Set-PSReadLineOption -ExtraPromptLineCount ($prompt.Split("`n").Length - 1) } >> perf.txt
 
         # Return the prompt
         $prompt
```

<details><summary>And got back this:</summary>

```pwsh
[32;1mDays              : [0m0
[32;1mHours             : [0m0
[32;1mMinutes           : [0m0
[32;1mSeconds           : [0m0
[32;1mMilliseconds      : [0m2
[32;1mTicks             : [0m27709
[32;1mTotalDays         : [0m3.20706018518519E-08
[32;1mTotalHours        : [0m7.69694444444444E-07
[32;1mTotalMinutes      : [0m4.61816666666667E-05
[32;1mTotalSeconds      : [0m0.0027709
[32;1mTotalMilliseconds : [0m2.7709



[32;1mDays              : [0m0
[32;1mHours             : [0m0
[32;1mMinutes           : [0m0
[32;1mSeconds           : [0m0
[32;1mMilliseconds      : [0m0
[32;1mTicks             : [0m1099
[32;1mTotalDays         : [0m1.27199074074074E-09
[32;1mTotalHours        : [0m3.05277777777778E-08
[32;1mTotalMinutes      : [0m1.83166666666667E-06
[32;1mTotalSeconds      : [0m0.0001099
[32;1mTotalMilliseconds : [0m0.1099



[32;1mDays              : [0m0
[32;1mHours             : [0m0
[32;1mMinutes           : [0m0
[32;1mSeconds           : [0m0
[32;1mMilliseconds      : [0m0
[32;1mTicks             : [0m998
[32;1mTotalDays         : [0m1.15509259259259E-09
[32;1mTotalHours        : [0m2.77222222222222E-08
[32;1mTotalMinutes      : [0m1.66333333333333E-06
[32;1mTotalSeconds      : [0m9.98E-05
[32;1mTotalMilliseconds : [0m0.0998



[32;1mDays              : [0m0
[32;1mHours             : [0m0
[32;1mMinutes           : [0m0
[32;1mSeconds           : [0m0
[32;1mMilliseconds      : [0m0
[32;1mTicks             : [0m803
[32;1mTotalDays         : [0m9.29398148148148E-10
[32;1mTotalHours        : [0m2.23055555555556E-08
[32;1mTotalMinutes      : [0m1.33833333333333E-06
[32;1mTotalSeconds      : [0m8.03E-05
[32;1mTotalMilliseconds : [0m0.0803



[32;1mDays              : [0m0
[32;1mHours             : [0m0
[32;1mMinutes           : [0m0
[32;1mSeconds           : [0m0
[32;1mMilliseconds      : [0m0
[32;1mTicks             : [0m974
[32;1mTotalDays         : [0m1.12731481481481E-09
[32;1mTotalHours        : [0m2.70555555555556E-08
[32;1mTotalMinutes      : [0m1.62333333333333E-06
[32;1mTotalSeconds      : [0m9.74E-05
[32;1mTotalMilliseconds : [0m0.0974



[32;1mDays              : [0m0
[32;1mHours             : [0m0
[32;1mMinutes           : [0m0
[32;1mSeconds           : [0m0
[32;1mMilliseconds      : [0m0
[32;1mTicks             : [0m743
[32;1mTotalDays         : [0m8.59953703703704E-10
[32;1mTotalHours        : [0m2.06388888888889E-08
[32;1mTotalMinutes      : [0m1.23833333333333E-06
[32;1mTotalSeconds      : [0m7.43E-05
[32;1mTotalMilliseconds : [0m0.0743



[32;1mDays              : [0m0
[32;1mHours             : [0m0
[32;1mMinutes           : [0m0
[32;1mSeconds           : [0m0
[32;1mMilliseconds      : [0m0
[32;1mTicks             : [0m797
[32;1mTotalDays         : [0m9.22453703703704E-10
[32;1mTotalHours        : [0m2.21388888888889E-08
[32;1mTotalMinutes      : [0m1.32833333333333E-06
[32;1mTotalSeconds      : [0m7.97E-05
[32;1mTotalMilliseconds : [0m0.0797



[32;1mDays              : [0m0
[32;1mHours             : [0m0
[32;1mMinutes           : [0m0
[32;1mSeconds           : [0m0
[32;1mMilliseconds      : [0m0
[32;1mTicks             : [0m881
[32;1mTotalDays         : [0m1.01967592592593E-09
[32;1mTotalHours        : [0m2.44722222222222E-08
[32;1mTotalMinutes      : [0m1.46833333333333E-06
[32;1mTotalSeconds      : [0m8.81E-05
[32;1mTotalMilliseconds : [0m0.0881



[32;1mDays              : [0m0
[32;1mHours             : [0m0
[32;1mMinutes           : [0m0
[32;1mSeconds           : [0m0
[32;1mMilliseconds      : [0m0
[32;1mTicks             : [0m931
[32;1mTotalDays         : [0m1.0775462962963E-09
[32;1mTotalHours        : [0m2.58611111111111E-08
[32;1mTotalMinutes      : [0m1.55166666666667E-06
[32;1mTotalSeconds      : [0m9.31E-05
[32;1mTotalMilliseconds : [0m0.0931



[32;1mDays              : [0m0
[32;1mHours             : [0m0
[32;1mMinutes           : [0m0
[32;1mSeconds           : [0m0
[32;1mMilliseconds      : [0m0
[32;1mTicks             : [0m696
[32;1mTotalDays         : [0m8.05555555555556E-10
[32;1mTotalHours        : [0m1.93333333333333E-08
[32;1mTotalMinutes      : [0m1.16E-06
[32;1mTotalSeconds      : [0m6.96E-05
[32;1mTotalMilliseconds : [0m0.0696



[32;1mDays              : [0m0
[32;1mHours             : [0m0
[32;1mMinutes           : [0m0
[32;1mSeconds           : [0m0
[32;1mMilliseconds      : [0m0
[32;1mTicks             : [0m617
[32;1mTotalDays         : [0m7.1412037037037E-10
[32;1mTotalHours        : [0m1.71388888888889E-08
[32;1mTotalMinutes      : [0m1.02833333333333E-06
[32;1mTotalSeconds      : [0m6.17E-05
[32;1mTotalMilliseconds : [0m0.0617


```
</details>

On a core i9-11900H. Not sure how big the difference is on slower CPUs.

Reply:
I get similar results (i7-8750H, AC plugged in).

But I still think we should avoid doing it on every prompt render. Does the count of lines really change every time?

Reply:
Only if someone defines a `format` that makes it so as far as I can tell.

Reply:
I can't think of a situation where a `format` definition would dynamically change that. 

For instance, I have a 2-line prompt (definition [here](https://github.com/rashil2000/dotfiles/blob/master/starship/starship.toml)). When the terminal's width is really small, text overflows into the third line. But the output of ```$prompt.Split("`n").Length - 1``` is still 2. Unchanged.

Reply:
@davidkna Also commented on the issue, what's your opinion?

Both will work obviously. It's just sticking a call to `global:prompt` at the start and doing the same thing just once instead of inside the `global:prompt` function.

Is a prompt that changes the number of lines it has dynamically something likely? (Not automatic line wrapping, but actual newline chars).

Reply:
Since the lack of this setting does not break anything (except for some formatting issue with third-party plugins like PSFzf), instead of trying to calculate the value, why don't we just tell the users to run the command themselves in the docs somewhere, and save the overhead altogether (whether it's once or on every render)?

Reply:
They are likely to miss setting it, the installation instructions for starship are expected to only include evaluating its init script.

Reply:
I meant somewhere in the Advanced Configuration section, since a majority of the users don't need it.

Reply:
![Hmm](https://user-images.githubusercontent.com/24731903/149665425-eb7c6e39-5f5b-47ff-8865-8d8c5395e282.png)

Reply:
With `--profile=` the prompt is only going to become more dynamic. Since the perf hit seems to be small, I'm fine with this approach for now.